frontend/src/metabase/admin/settings/components/SettingsBatchForm.jsx

```diff
@@ -0,0 +1,328 @@
+          valid = false;
+        }
+
+        if (element.validations) {
+          element.validations.forEach(function(validation) {
+            validationErrors[element.key] = this.validateElement(
+              validation,
+              formData[element.key],
+              element,
+            );
+            if (validationErrors[element.key]) {
+              valid = false;
+            }
+          }, this);
+        }
+      }, this);
+    }
+
+    if (
+      this.state.valid !== valid ||
+      !_.isEqual(this.state.validationErrors, validationErrors)
+    ) {
+      this.setState({ valid, validationErrors });
+    }
+  }
+
+  handleChangeEvent(key, value) {
+    this.setState(previousState => ({
+      dirty: true,
+      formData: {
+        ...previousState.formData,
+        [key]: MetabaseUtils.isEmpty(value) ? null : value,
+      },
+    }));
+  }
+
+  handleFormErrors(error) {
+    // parse and format
+    let formErrors = {};
+    if (error.data && error.data.message) {
+      formErrors.message = error.data.message;
+    } else {
+      formErrors.message = t`Looks like we ran into some problems`;
+    }
+
+    if (error.data && error.data.errors) {
+      formErrors.elements = error.data.errors;
+    }
+
+    return formErrors;
+  }
+
+  updateSettings = e => {
+    e.preventDefault();
+
+    let { formData, valid } = this.state;
+
+    if (valid) {
+      this.setState({
+        formErrors: null,
+        submitting: "working",
+      });
+
+      this.props.updateSettings(formData).then(
+        () => {
+          this.setState({ dirty: false, submitting: "success" });
+
+          // show a confirmation for 3 seconds, then return to normal
+          setTimeout(() => this.setState({ submitting: "default" }), 3000);
+        },
+        error => {
+          this.setState({
+            submitting: "default",
+            formErrors: this.handleFormErrors(error),
+          });
+        },
+      );
+    }
+  };
+
+  render() {
+    const { elements, settingValues } = this.props;
+    const {
+      formData,
+      formErrors,
+      submitting,
+      valid,
+      dirty,
+      validationErrors,
+    } = this.state;
+
+    const layout = this.props.layout || [
+      { settings: elements.map(element => element.key) },
+    ];
+    const settings = _.indexBy(elements, "key");
+
+    const getSetting = key => {
+      const element = settings[key];
+      if (!element) {
+        console.warn("Missing setting", element);
+        return null;
+      }
+      // merge together data from a couple places to provide a complete view of the Element state
+      let errorMessage =
+        formErrors && formErrors.elements
+          ? formErrors.elements[element.key]
+          : validationErrors[element.key];
+      let value =
+        formData[element.key] == null
+          ? element.defaultValue
+          : formData[element.key];
+
+      return (
+        <SettingsSetting
```

Comment:
The name of this component has always wigged me out a little. Maybe just `Setting`?